components/rollup_node/block_gen.md

```diff
@@ -0,0 +1,147 @@
+# Block Generation
+
+The logic which is used to generate the rollup chain from L1.
+
+## Glossary
+
+- **Block Inputs**: All information required to generate a full Ethereum block, given the pre-state and previous block. For example, does NOT include a state root for that block's input.
+- **Rollup Blockchain**: The rollup blocks generated by performing the rollup's state transition function over all block inputs. For example, DOES include a state root for that block's input.
+
+## Summary
+
+The rollup chain can be deterministically generated given an L1 Ethereum chain. The fact that the entire rollup chain can be generated based on L1 blocks is _what makes OE a rollup_. This process can be represented as:
+
+```python
+f(l1_blockchain) -> rollup_blockchain
+```
+
+In this document we define a block generation function which is designed to be:
+
+1. Context independent given a constant sized subset of the L1 chain.
+2. Support sequencers and sequencer consensus.
+3. Resilient sequencer censorship.
+
+## Rollup Epochs
+
+The rollup chain is subdivided into epochs. There is a 1:1 correspondence between L1 block numbers and epoch numbers. For L1 block number `n`, there is a corresponding rollup epoch `n` which can only be generated after L1 block number `n` is added to the L1 chain. Epoch contain one to many rollup blocks.
+
+### Types of Blocks within Epochs
+
+Within these epochs, there are two block types:
+
+1. Deposit block
+2. Sequencer block
+
+Deposit blocks contain contextual information about L1 such as the block hash. They also contain transactions initiated on L1 by any user or contract.
+
+Sequencer blocks are submitted by the sequencer and target _future_ epochs which satisfy the following two conditions:
+
+1. Target epoch is larger than the current L1 block number.
+2. Target epoch is less than the current L1 block number PLUS `sequencing_window_size`.
+
+The ability for the sequencer to append sequencer blocks to future epochs allows the sequencer to predict and influence the state of the epoch before the L1 chain has mined it. This property is what enables _fast transaction confirmations_ via the sequencer replicator.
```

Comment:
> Is it necessary for the sequencer to pick an epoch number in order to promptly confirm the the transaction?

If by "promptly confirm" you mean both confirm that the transaction will be included **and** _compute the result of the transaction_ then yes. Of these two, the more important property is the latter property of predicting the future state. It is critical that the sequencer can predict future state because 1) the sequencer can tell users what the chain will look like in the future with low latency, 2) the sequencer can be confident the user has a high enough balance to pay their fee.

The intuition for why this is possible is because the rollup chain is made up of deposit blocks & sequencer blocks. The sequencer knows the deposit block for a future epoch `sequencer_window_size` blocks in advance. And then the sequencer is able to themselves determine what sequencer blocks to submit. So even before they've submitted their blocks on chain, they know that assuming they are able to submit them in time they can pre-compute the state of L2 even before L1 reflects it.